Suppress repetitive MessageBoxes

roberthudson opened the pull request:
This addresses @tridge 's request to show "do this for all"
#2519
![image](https://user-images.githubusercontent.com/21204270/109434710-8ffa9300-79e4-11eb-86c1-02ba830b6a71.png)

The following approach is used:
There are 2 static classes named CustomMessageBox; one in the  MissionPlanner.MsgBox namespace and one in System. Both need adjusting, as one calls the other.

-The System class has a bunch of overrides. I added an optional parameter DoThisForAllText which, when set, will display a checkbox for example "Do this for all current items"
-The MissionPlanner.MsgBox class, which actually renders the messagebox will add the checkbox to the bottom of the form.

If the checkbox is selected, it will store the selection in a static bool property called DoThisForAll
(I understand this may not be the best approach, but it works. I'm open to ideas how to better pass back the checkbox selection)

in ConfigRawParams.cs I added some bools to track whether the dialog should be suppressed or not. If so, the last choice (Yes/No) is saved and used instead of showing the dialog.

tridge commented:
nice! testing now

tridge commented:
does this need to be enabled on the various dialogs?
I see this:
![image](https://user-images.githubusercontent.com/831867/109453045-2e4e2d80-7aa5-11eb-857e-043f98074a5f.png)

tridge commented:
here are some example param files for fixed wing that trigger this:
http://uav.tridgell.net/tmp/testparam/
just start sitl for a plane in MP and alternately load those two param files, it gives the dialog for out of range on the RLL2SRV params

roberthudson commented:
Sorry! I somehow missed adding the custom dialog to the Full Parameter Tree. This is fixed in the March 1 commit.
While I was at it, I also added the ReadOnly check to Full Parameter Tree - it was missing.

tridge commented:
I've tested this and it works really well now, thanks!

tridge commented:
just found a small issue. After I finish loading the params when I have selected the "for all" checkbox and I then go to load a different parameter file I don't get any warning dialogs at all. I think the correct behaviour would be that the state would clear after the end of the current operation

roberthudson commented:
Several fixes in this commit....

1.  if there are several missing params, the messagebox wraps and limits to (default of)  25
![image](https://user-images.githubusercontent.com/21204270/109573465-1f737500-7abd-11eb-89dc-5282a2b8ab37.png)

It used to be this ridiculous mile-wide beast
![image](https://user-images.githubusercontent.com/21204270/109574162-74af8680-7abd-11eb-832e-6c9a3db2753a.png)


2. DoThisForAll flag resets when a new config file is loaded

3. Fixes a null error when the param grid contains an empty cell. In some cases row.Cells[1].Value could be null. 
changed
`if (row.Cells[1].Value.ToString() != value)`
to
`if (row.Cells[1].Value != null && row.Cells[1].Value.ToString() != value)`

(Thanks for spotting that tridge, I'm surprised this wasn't detected long ago)

4. Added a few try/catches

roberthudson commented:
> yeah, having a static seems wrong here, and just prone to issues.

True, I realize it's not ideal. But what's the alternative?

Yeah it's a static, but it's just a simple boolean flag. There isn't really a thread-safety issue. And the flag gets cleared each time the messagebox is displayed.

My first idea was adding some more enums to the DialogResut (adding OKToAll, YesToAll, NoToAll)

`public enum DialogResult
        {
            None,
            OK,
            Cancel,
            Abort,
            Retry,
            Ignore,
            Yes,
            No,
            OKToAll,
            YesToAll,
            NoToAll
        }`

But that won't work. 
The MissionPlanner.MsgBox.CustomMessageBox returns a System enum which of course is sealed.

What do you suggest? How else is the calling function going to know the user checked "Do this for all"?

tridge commented:
> What do you suggest? How else is the calling function going to know the user checked "Do this for all"?

I'm not a C# dev, so this may not work, but perhaps CustomMessageBox could be subclassed to CustomMessageBoxCheckAll and the functionality and static be only in the subclass. Then we could use the subclass just in the places it makes sense, such as the param interface, without risking unexpected impact on other use cases.
There seems to be over 900 uses of CustomMessageBox, and I think it would be easier to opt-in the cases that matter

tridge commented on Common.cs:
```diff
@@ -237,5 +238,19 @@ static void chk_CheckStateChanged(object sender, EventArgs e)
+            {
+                list += item + "\n";
+            }
+            if (missing.Count > 25)
```
i think the 25 here should be maxItems

roberthudson replied:
See https://github.com/ArduPilot/MissionPlanner/issues/2519

tridge commented on Common.cs:
```diff
@@ -237,5 +238,19 @@ static void chk_CheckStateChanged(object sender, EventArgs e)
+                list += item + "\n";
+            }
+            if (missing.Count > 25)
+                list += "(and " + (missing.Count - 25) + " more).";
```
and here

roberthudson replied:
Yeah right. I added that param very last minute - it should be maxItems

roberthudson replied:
See https://github.com/ArduPilot/MissionPlanner/issues/2519

meee1 commented:
i do agree with tridge, it should be opt-in. as the dialog itself is used everywhere, and unrelated to most likely 850 of those 900 usages

roberthudson commented:
Ok, I'm getting to know Mission Planner more and more.
So.....  I'm rethinking this issue yet again. We're _**still**_ on the wrong approach.

Preface at @meee1: No disrespect in the following. Mission Planner is just amazing and you've done an incredible job with it. But with system this complex, sometimes small details get overlooked. 
I've been asked to look at one particular issue (the Whack-a Mole dialog game) - so that's where I'm focusing. 

There's a bit of a logic flaw. Here's why we're getting all those dialog boxes. They popup on the CellValueChanged event. 
This event fires for each row when the grid is loaded. What we need is a bool flag "IsLoading" that is set to true when the grid is populated. Then clear the flag.
While IsLoading=true we want to ignore if the value is ReadOnly. Who cares? We only care IF the user tries to CHANGE it.

Here's what I suggest:
-set an IsLoading flag and suppress the MessageBox while the grid is populating. Then clear the flag.

-~~then, if the user tries to change a ReadOnly cell we show the dialog.~~ Lock the cell so they can't change it. Mark the cell in orange or something. 

-we can scrap the Do This For All checkbox. It's not needed

-for the Out Of Range cells, mark them in some other color as a visual cue. Maybe yellow

-after populating the grid display a dialog "n values are out of range. They have been highlighted yellow"
